File: src/terminal.rs
use std::sync::Mutex;

use crate::MainWindow;

pub const GRID_COLS: usize = 80;
pub const GRID_ROWS: usize = 22;

pub struct Screen {
    cells: [u8; GRID_COLS * GRID_ROWS],
    row: usize,
    col: usize,
    esc: bool,
    csi: bool,
    csi_params: Vec<u8>,
}

impl Screen {
    pub fn new() -> Self {
        Self {
            cells: [b' '; GRID_COLS * GRID_ROWS],
            row: 0,
            col: 0,
            esc: false,
            csi: false,
            csi_params: Vec::new(),
        }
    }

    pub fn feed(&mut self, data: &[u8]) {
        for &b in data {
            self.step(b);
        }
    }

    fn step(&mut self, b: u8) {
        if self.csi {
            if (0x40..=0x7e).contains(&b) {
                self.csi = false;
                let params = std::mem::take(&mut self.csi_params);
                self.csi_exec(&params, b);
            } else {
                self.csi_params.push(b);
            }
            return;
        }

        if self.esc {
            self.esc = false;
            if b == b'[' {
                self.csi = true;
                self.csi_params.clear();
            }
            return;
        }

        match b {
            0x1b => self.esc = true,
            b'\r' => self.col = 0,
            b'\n' => self.newline(),
            0x08 => {
                if self.col > 0 {
                    self.col -= 1;
                }
            }
            b'\t' => {
                self.col = (self.col / 8 + 1) * 8;
                if self.col >= GRID_COLS {
                    self.col = GRID_COLS - 1;
                }
            }
            _ => {
                if b >= 0x20 {
                    self.put(b);
                }
            }
        }
    }

    fn put(&mut self, b: u8) {
        if self.row < GRID_ROWS {
            self.cells[self.row * GRID_COLS + self.col] = b;
        }
        self.col += 1;
        if self.col >= GRID_COLS {
            self.col = 0;
            self.newline();
        }
    }

    fn newline(&mut self) {
        if self.row + 1 >= GRID_ROWS {
            self.scroll_up();
        } else {
            self.row += 1;
        }
    }

    fn scroll_up(&mut self) {
        self.cells.copy_within(GRID_COLS..GRID_COLS * GRID_ROWS, 0);
        self.cells[GRID_COLS * (GRID_ROWS - 1)..GRID_COLS * GRID_ROWS].fill(b' ');
    }

    fn erase_line(&mut self, mode: i32) {
        let row_start = self.row * GRID_COLS;
        match mode {
            0 => {
                self.cells[row_start + self.col..row_start + GRID_COLS].fill(b' ');
            }
            1 => {
                self.cells[row_start..=(row_start + self.col).min(row_start + GRID_COLS - 1)]
                    .fill(b' ');
            }
            _ => {
                self.cells[row_start..row_start + GRID_COLS].fill(b' ');
            }
        }
    }

    fn erase_screen(&mut self, mode: i32) {
        match mode {
            0 => {
                self.erase_line(0);
                for r in (self.row + 1)..GRID_ROWS {
                    self.cells[r * GRID_COLS..(r + 1) * GRID_COLS].fill(b' ');
                }
            }
            1 => {
                self.erase_line(1);
                for r in 0..self.row {
                    self.cells[r * GRID_COLS..(r + 1) * GRID_COLS].fill(b' ');
                }
            }
            _ => {
                self.cells.fill(b' ');
            }
        }
    }

    fn csi_exec(&mut self, params: &[u8], final_byte: u8) {
        if matches!(
            final_byte,
            b'm' | b'h' | b'l' | b'r' | b's' | b'u' | b't' | b'Z'
        ) {
            return;
        }

        let p = parse_params(params);
        let get = |i: usize| -> usize {
            if i < p.len() {
                p[i].max(1) as usize
            } else {
                1
            }
        };

        match final_byte {
            b'H' | b'f' => {
                self.row = get(0).saturating_sub(1).min(GRID_ROWS - 1);
                self.col = get(1).saturating_sub(1).min(GRID_COLS - 1);
            }
            b'A' => {
                self.row = self.row.saturating_sub(get(0));
            }
            b'B' => {
                self.row = (self.row + get(0)).min(GRID_ROWS - 1);
            }
            b'C' => {
                self.col = (self.col + get(0)).min(GRID_COLS - 1);
            }
            b'D' => {
                self.col = self.col.saturating_sub(get(0));
            }
            b'G' => {
                self.col = get(0).saturating_sub(1).min(GRID_COLS - 1);
            }
            b'd' => {
                self.row = get(0).saturating_sub(1).min(GRID_ROWS - 1);
            }
            b'J' => {
                self.erase_screen(p.first().copied().unwrap_or(0));
            }
            b'K' => {
                self.erase_line(p.first().copied().unwrap_or(0));
            }
            _ => {}
        }
    }

    pub fn as_text(&self) -> String {
        let mut s = String::with_capacity(GRID_COLS * GRID_ROWS + GRID_ROWS);
        for r in 0..GRID_ROWS {
            if r > 0 {
                s.push('\n');
            }
            let row_data = &self.cells[r * GRID_COLS..(r + 1) * GRID_COLS];
            s.push_str(&String::from_utf8_lossy(row_data));
        }
        s
    }
}

fn parse_params(data: &[u8]) -> Vec<i32> {
    let mut out = Vec::new();
    let mut cur = 0;
    let mut has_cur = false;

    for &b in data {
        if b.is_ascii_digit() {
            cur = cur * 10 + (b - b'0') as i32;
            has_cur = true;
        } else if b == b';' {
            out.push(if has_cur { cur } else { 1 });
            cur = 0;
            has_cur = false;
        }
    }

    if has_cur {
        out.push(cur);
    } else if out.is_empty() {
        out.push(1);
    }
    out
}

struct Session {
    master_fd: i32,
    process_group: libc::pid_t,
}

pub struct TerminalManager {
    session: Mutex<Option<Session>>,
    reader_thread: Mutex<Option<std::thread::JoinHandle<()>>>,
}

impl TerminalManager {
    pub fn new() -> Self {
        Self {
            session: Mutex::new(None),
            reader_thread: Mutex::new(None),
        }
    }

    #[allow(dead_code)]
    pub fn is_running(&self) -> bool {
        self.session.lock().unwrap().is_some()
    }

    pub fn spawn(&self, exec: &str, setting_id: i32, weak_window: slint::Weak<MainWindow>) {
        let mut session_lock = self.session.lock().unwrap();
        if session_lock.is_some() {
            eprintln!("terminal already running");
            return;
        }

        let mut master_fd: libc::c_int = -1;
        let mut slave_fd: libc::c_int = -1;
        let mut ws = libc::winsize {
            ws_row: GRID_ROWS as u16,
            ws_col: GRID_COLS as u16,
            ws_xpixel: 0,
            ws_ypixel: 0,
        };

        let res = unsafe {
            libc::openpty(
                &mut master_fd,
                &mut slave_fd,
                std::ptr::null_mut(),
                std::ptr::null_mut(),
                &mut ws,
            )
        };

        if res < 0 {
            eprintln!("openpty error: errno {}", std::io::Error::last_os_error());
            let weak = weak_window.clone();
            slint::invoke_from_event_loop(move || {
                if let Some(w) = weak.upgrade() {
                    w.set_terminal_active(false);
                    w.set_terminal_id(0);
                }
            })
            .ok();
            return;
        }

        let child = unsafe { libc::fork() };
        if child < 0 {
            eprintln!("fork error: errno {}", std::io::Error::last_os_error());
            unsafe {
                libc::close(master_fd);
                libc::close(slave_fd);
            }
            return;
        }

        if child == 0 {
            unsafe {
                libc::close(master_fd);
                libc::setsid();
                libc::ioctl(slave_fd, libc::TIOCSCTTY, 0);
                libc::dup2(slave_fd, libc::STDIN_FILENO);
                libc::dup2(slave_fd, libc::STDOUT_FILENO);
                libc::dup2(slave_fd, libc::STDERR_FILENO);
                if slave_fd > 2 {
                    libc::close(slave_fd);
                }

                let sh = std::ffi::CString::new("sh").unwrap();
                let c = std::ffi::CString::new("-c").unwrap();
                let cmd = std::ffi::CString::new(exec).unwrap();
                libc::execlp(
                    sh.as_ptr(),
                    sh.as_ptr(),
                    c.as_ptr(),
                    cmd.as_ptr(),
                    std::ptr::null::<libc::c_char>(),
                );
                libc::_exit(127);
            }
        }

        unsafe {
            libc::close(slave_fd);
            libc::setpgid(child, child);
        }

        *session_lock = Some(Session {
            master_fd,
            process_group: child,
        });

        let weak_ui = weak_window.clone();
        slint::invoke_from_event_loop(move || {
            if let Some(w) = weak_ui.upgrade() {
                w.set_terminal_active(true);
                w.set_terminal_id(setting_id);
            }
        })
        .ok();

        let weak_reader = weak_window.clone();
        let handle = std::thread::spawn(move || {
            let mut screen = Screen::new();
            let mut buffer = [0u8; 4096];
            let mut last_text = String::new();

            loop {
                let n = unsafe {
                    libc::read(
                        master_fd,
                        buffer.as_mut_ptr() as *mut libc::c_void,
                        buffer.len(),
                    )
                };
                if n <= 0 {
                    break;
                }

                screen.feed(&buffer[..n as usize]);
                let text = screen.as_text();
                if text != last_text {
                    last_text = text.clone();
                    let weak = weak_reader.clone();
                    let text_shared: slint::SharedString = text.into();
                    slint::invoke_from_event_loop(move || {
                        if let Some(w) = weak.upgrade() {
                            w.set_terminal_output(text_shared);
                        }
                    })
                    .ok();
                }
            }

            let mut status: libc::c_int = 0;
            unsafe {
                libc::waitpid(child, &mut status, 0);
            }
            eprintln!("interactive session ended");

            let weak = weak_reader.clone();
            slint::invoke_from_event_loop(move || {
                if let Some(w) = weak.upgrade() {
                    w.set_terminal_active(false);
                    w.set_terminal_id(0);
                    w.set_terminal_output("".into());
                }
            })
            .ok();
        });

        *self.reader_thread.lock().unwrap() = Some(handle);
    }

    pub fn write(&self, text: &str) {
        let lock = self.session.lock().unwrap();
        if let Some(ref session) = *lock {
            unsafe {
                libc::write(
                    session.master_fd,
                    text.as_ptr() as *const libc::c_void,
                    text.len(),
                );
                libc::tcdrain(session.master_fd);
            }
        }
    }

    pub fn terminate(&self, weak_window: &slint::Weak<MainWindow>) {
        let session = self.session.lock().unwrap().take();
        if let Some(session) = session {
            eprintln!("terminating interactive program");
            unsafe {
                libc::kill(-session.process_group, libc::SIGTERM);
            }
            std::thread::sleep(std::time::Duration::from_millis(200));
            unsafe {
                libc::kill(-session.process_group, libc::SIGKILL);
                libc::close(session.master_fd);
            }
        } else {
            eprintln!("requested to terminate interactive program but none is running");
        }

        let handle = self.reader_thread.lock().unwrap().take();
        if let Some(handle) = handle {
            let _ = handle.join();
        }

        let weak = weak_window.clone();
        slint::invoke_from_event_loop(move || {
            if let Some(w) = weak.upgrade() {
                w.set_terminal_active(false);
                w.set_terminal_id(0);
                w.set_terminal_output("".into());
            }
        })
        .ok();
    }

    pub fn terminate_all(&self) {
        let session = self.session.lock().unwrap().take();
        if let Some(session) = session {
            unsafe {
                libc::kill(-session.process_group, libc::SIGKILL);
                libc::close(session.master_fd);
            }
        }

        let handle = self.reader_thread.lock().unwrap().take();
        if let Some(handle) = handle {
            let _ = handle.join();
        }
    }
}

impl Drop for TerminalManager {
    fn drop(&mut self) {
        self.terminate_all();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_screen_basic() {
        let mut screen = Screen::new();
        screen.feed(b"Hello World");
        let text = screen.as_text();
        assert!(text.starts_with("Hello World"));
    }

    #[test]
    fn test_screen_cursor_and_newlines() {
        let mut screen = Screen::new();
        screen.feed(b"Line 1\r\nLine 2");
        let text = screen.as_text();
        let lines: Vec<&str> = text.lines().collect();
        assert!(lines[0].starts_with("Line 1"));
        assert!(lines[1].starts_with("Line 2"));
    }

    #[test]
    fn test_screen_ansi_cursor() {
        let mut screen = Screen::new();
        // CSI 2;5H -> move to row 2 (index 1), col 5 (index 4)
        screen.feed(b"\x1b[2;5HTest");
        let text = screen.as_text();
        let lines: Vec<&str> = text.lines().collect();
        assert_eq!(&lines[1][4..8], "Test");
    }
}

File: src/terminal.cpp
#include "terminal.h"

#include <algorithm>
#include <atomic>
#include <cerrno>
#include <cstring>
#include <format>
#include <memory>
#include <print>
#include <thread>

#include <signal.h>
#include <sys/ioctl.h>
#include <sys/wait.h>
#include <termios.h>
#include <unistd.h>
#include <utmp.h>

#include <pty.h>

#include "unique_fd.h"
#include "nothrow.h"

static std::string errno_str() {
    char buf[256];
    if (strerror_r(errno, buf, sizeof(buf)) == 0) {
        return std::string(buf);
    }
    return std::format("errno {}", errno);
}

Screen::Screen()
    : row_(0), col_(0), esc_(false) {
    std::ranges::fill(cells_, static_cast<uint8_t>(' '));
}

void Screen::feed(const uint8_t* data, size_t len) {
    for (size_t i = 0; i < len; ++i) {
        step(data[i]);
    }
}

void Screen::step(uint8_t b) {
    if (!csi_params_.empty()) {
        if (b >= 0x40 && b <= 0x7e) {
            auto params = std::move(csi_params_);
            csi_params_.clear();
            csi_exec(params.data(), params.size(), b);
        } else {
            csi_params_.push_back(b);
        }
        return;
    }

    if (esc_) {
        esc_ = false;
        if (b == '[') {
            csi_params_.clear();
        }
        return;
    }

    switch (b) {
        case 0x1b:
            esc_ = true;
            break;
        case '\r':
            col_ = 0;
            break;
        case '\n':
            newline();
            break;
        case '\x08':
            if (col_ > 0) col_--;
            break;
        case '\t':
            col_ = (col_ / 8 + 1) * 8;
            if (col_ >= GRID_COLS) col_ = GRID_COLS - 1;
            break;
        default:
            if (b >= 0x20) {
                put(b);
            }
            break;
    }
}

void Screen::put(uint8_t b) {
    if (row_ < GRID_ROWS) {
        cells_[idx(row_, col_)] = b;
    }
    col_++;
    if (col_ >= GRID_COLS) {
        col_ = 0;
        newline();
    }
}

void Screen::newline() {
    if (row_ + 1 >= GRID_ROWS) {
        scroll_up();
    } else {
        row_++;
    }
}

void Screen::scroll_up() {
    std::ranges::copy(cells_ + GRID_COLS, cells_ + GRID_COLS * GRID_ROWS, cells_);
    std::ranges::fill(cells_ + GRID_COLS * (GRID_ROWS - 1),
                      cells_ + GRID_COLS * GRID_ROWS,
                      static_cast<uint8_t>(' '));
}

void Screen::erase_line(int mode) {
    int row_start = row_ * GRID_COLS;
    switch (mode) {
        case 0:
            std::ranges::fill(cells_ + row_start + col_, cells_ + row_start + GRID_COLS, ' ');
            break;
        case 1:
            std::ranges::fill(cells_ + row_start, cells_ + row_start + col_ + 1, ' ');
            break;
        default:
            std::ranges::fill(cells_ + row_start, cells_ + row_start + GRID_COLS, ' ');
            break;
    }
}

void Screen::erase_screen(int mode) {
    switch (mode) {
        case 0:
            erase_line(0);
            for (int r = row_ + 1; r < GRID_ROWS; r++) {
                std::ranges::fill(cells_ + r * GRID_COLS,
                                  cells_ + (r + 1) * GRID_COLS,
                                  static_cast<uint8_t>(' '));
            }
            break;
        case 1:
            erase_line(1);
            for (int r = 0; r < row_; r++) {
                std::ranges::fill(cells_ + r * GRID_COLS,
                                  cells_ + (r + 1) * GRID_COLS,
                                  static_cast<uint8_t>(' '));
            }
            break;
        default:
            std::ranges::fill(cells_, cells_ + sizeof(cells_),
                              static_cast<uint8_t>(' '));
            break;
    }
}

static std::vector<int> parse_params(const uint8_t* data, size_t len) {
    std::vector<int> out;
    int cur = 0;
    bool has_cur = false;

    for (size_t i = 0; i < len; i++) {
        uint8_t b = data[i];
        if (b >= '0' && b <= '9') {
            cur = cur * 10 + (b - '0');
            has_cur = true;
        } else if (b == ';') {
            out.push_back(has_cur ? cur : 1);
            cur = 0;
            has_cur = false;
        }
    }

    if (has_cur) {
        out.push_back(cur);
    } else if (out.empty()) {
        out.push_back(1);
    }
    return out;
}

void Screen::csi_exec(const uint8_t* params, size_t len, uint8_t final_byte) {
    if (final_byte == 'm' || final_byte == 'h' || final_byte == 'l' ||
        final_byte == 'r' || final_byte == 's' || final_byte == 'u' ||
        final_byte == 't' || final_byte == 'Z') {
        return;
    }

    auto p = parse_params(params, len);
    auto get = [&](size_t i) -> int {
        return (i < p.size()) ? p[i] : 1;
    };

    switch (final_byte) {
        case 'H':
        case 'f':
            row_ = std::clamp(get(0) - 1, 0, GRID_ROWS - 1);
            col_ = std::clamp(get(1) - 1, 0, GRID_COLS - 1);
            break;
        case 'A':
            row_ = std::max(0, row_ - get(0));
            break;
        case 'B':
            row_ = std::min(GRID_ROWS - 1, row_ + get(0));
            break;
        case 'C':
            col_ = std::min(GRID_COLS - 1, col_ + get(0));
            break;
        case 'D':
            col_ = std::max(0, col_ - get(0));
            break;
        case 'G':
            col_ = std::clamp(get(0) - 1, 0, GRID_COLS - 1);
            break;
        case 'd':
            row_ = std::clamp(get(0) - 1, 0, GRID_ROWS - 1);
            break;
        case 'J':
            erase_screen(p.empty() ? 0 : p[0]);
            break;
        case 'K':
            erase_line(p.empty() ? 0 : p[0]);
            break;
        default:
            break;
    }
}

std::string Screen::as_text() const {
    std::string s;
    s.reserve(GRID_COLS * GRID_ROWS + GRID_ROWS);
    for (int r = 0; r < GRID_ROWS; r++) {
        if (r > 0) s.push_back('\n');
        const uint8_t* row_data = cells_ + r * GRID_COLS;
        s.append(reinterpret_cast<const char*>(row_data), GRID_COLS);
    }
    return s;
}

// --- TerminalManager implementation ---

TerminalManager::TerminalManager() = default;

TerminalManager::~TerminalManager() {
    terminate_all();
}

bool TerminalManager::is_running() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return session_ != nullptr;
}

void TerminalManager::spawn(const std::string& exec, int32_t setting_id,
                            slint::ComponentWeakHandle<MainWindow> weak_window) {
    std::lock_guard<std::mutex> lock(mutex_);

    if (session_) {
        std::println(std::cerr, "terminal already running");
        return;
    }

    int raw_master_fd = -1;
    pid_t child_pid = -1;

    struct winsize ws = {};
    ws.ws_row = GRID_ROWS;
    ws.ws_col = GRID_COLS;

    child_pid = forkpty(&raw_master_fd, nullptr, nullptr, &ws);
    if (child_pid < 0) {
        std::println(std::cerr, "forkpty error: {}", errno_str());
        slint::invoke_from_event_loop([weak_window, setting_id]() {
            if (auto w = weak_window.lock()) {
                (*w)->set_terminal_active(false);
                (*w)->set_terminal_id(0);
            }
        });
        return;
    }

    if (child_pid == 0) {
        setsid();
        execlp("sh", "sh", "-c", exec.c_str(), nullptr);
        std::println(std::cerr, "execlp: {}", errno_str());
        _exit(127);
    }

    UniqueFd master_fd(raw_master_fd);

    pid_t pgid = child_pid;
    setpgid(child_pid, pgid);

    session_ = make_nothrow<Session>(Session{
        .master_fd = master_fd.get(),
        .child_pid = child_pid,
        .process_group = pgid,
    });

    slint::invoke_from_event_loop([weak_window, setting_id]() {
        if (auto w = weak_window.lock()) {
            (*w)->set_terminal_active(true);
            (*w)->set_terminal_id(setting_id);
        }
    });

    int fd_for_reader = master_fd.get();
    reader_thread_ = std::thread([this, fd_for_reader, child_pid, weak_window]() {
        Screen screen;
        char buffer[4096];
        std::string last_text;

        while (true) {
            ssize_t n = ::read(fd_for_reader, buffer, sizeof(buffer));
            if (n <= 0) break;

            screen.feed(reinterpret_cast<uint8_t*>(buffer), n);
            std::string text = screen.as_text();
            if (text != last_text) {
                last_text = text;
                auto text_copy = text;
                slint::invoke_from_event_loop([weak_window, text_copy]() {
                    if (auto w = weak_window.lock()) {
                        (*w)->set_terminal_output(slint::SharedString(text_copy));
                    }
                });
            }
        }

        int status;
        waitpid(child_pid, &status, 0);
        std::println(std::cerr, "interactive session ended");

        slint::invoke_from_event_loop([weak_window]() {
            if (auto w = weak_window.lock()) {
                (*w)->set_terminal_active(false);
                (*w)->set_terminal_id(0);
                (*w)->set_terminal_output(slint::SharedString(""));
            }
        });
    });

    static_cast<void>(master_fd.release());
}

void TerminalManager::write(const std::string& text) {
    std::lock_guard<std::mutex> lock(mutex_);
    if (session_) {
        ::write(session_->master_fd, text.data(), text.size());
        tcdrain(session_->master_fd);
    }
}

void TerminalManager::terminate(slint::ComponentWeakHandle<MainWindow> weak_window) {
    std::unique_ptr<Session> session;
    {
        std::lock_guard<std::mutex> lock(mutex_);
        session = std::move(session_);
    }

    if (session) {
        std::println(std::cerr, "terminating interactive program");
        ::kill(-session->process_group, SIGTERM);
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
        ::kill(-session->process_group, SIGKILL);
        ::close(session->master_fd);
    } else {
        std::println(std::cerr, "requested to terminate interactive program but none is running");
    }

    if (reader_thread_.joinable()) {
        reader_thread_.join();
    }

    finish(weak_window);
}

void TerminalManager::terminate_all() {
    std::unique_ptr<Session> session;
    {
        std::lock_guard<std::mutex> lock(mutex_);
        session = std::move(session_);
    }

    if (session) {
        ::kill(-session->process_group, SIGKILL);
        ::close(session->master_fd);
    }

    if (reader_thread_.joinable()) {
        reader_thread_.join();
    }
}

void TerminalManager::finish(slint::ComponentWeakHandle<MainWindow> weak_window) {
    slint::invoke_from_event_loop([weak_window]() {
        if (auto w = weak_window.lock()) {
            (*w)->set_terminal_active(false);
            (*w)->set_terminal_id(0);
            (*w)->set_terminal_output(slint::SharedString(""));
        }
    });
}
